google/cloud/bigquery_storage_v1beta2/writer.py

```diff
@@ -0,0 +1,395 @@
+
+        Args:
+            reason:
+                The reason to close the stream. If ``None``, this is
+                considered an "intentional" shutdown.
+        """
+        with self._closing:
+            if self._closed:
+                return
+
+            # Stop consuming messages.
+            if self.is_active:
+                _LOGGER.debug("Stopping consumer.")
+                self._consumer.stop()
+            self._consumer = None
+
+            self._rpc.close()
+            self._rpc = None
+            self._closed = True
+            _LOGGER.debug("Finished stopping manager.")
+
+            # We know that no new items will be added to the queue because
+            # we've marked the stream as closed.
+            while not self._futures_queue.empty():
+                # Mark each future as failed. Since the consumer thread has
+                # stopped (or at least is attempting to stop), we won't get
+                # response callbacks to populate the remaining futures.
+                future = self._futures_queue.get_nowait()
+                if reason is None:
+                    exc = bqstorage_exceptions.StreamClosedError(
+                        "Stream closed before receiving a response."
+                    )
+                else:
+                    exc = reason
+                future.set_exception(exc)
+
+            for callback in self._close_callbacks:
+                callback(self, reason)
+
+    def _on_rpc_done(self, future):
+        """Triggered whenever the underlying RPC terminates without recovery.
+
+        This is typically triggered from one of two threads: the background
+        consumer thread (when calling ``recv()`` produces a non-recoverable
+        error) or the grpc management thread (when cancelling the RPC).
+
+        This method is *non-blocking*. It will start another thread to deal
+        with shutting everything down. This is to prevent blocking in the
+        background consumer and preventing it from being ``joined()``.
+        """
+        _LOGGER.info("RPC termination has signaled streaming pull manager shutdown.")
+        error = _wrap_as_exception(future)
+        thread = threading.Thread(
+            name=_RPC_ERROR_THREAD_NAME, target=self._shutdown, kwargs={"reason": error}
+        )
+        thread.daemon = True
+        thread.start()
+
+
+class AppendRowsFuture(polling_future.PollingFuture):
+    """Encapsulation of the asynchronous execution of an action.
+
+    This object is returned from long-running BigQuery Storage API calls, and
+    is the interface to determine the status of those calls.
+
+    This object should not be created directly, but is returned by other
+    methods in this library.
+    """
+
+    def __init__(self, manager: AppendRowsStream):
+        super().__init__()
+        self.__manager = manager
+        self.__cancelled = False
+        self._is_done = False
+
+    def cancel(self):
```

Comment:
cancel() on an individual future kills all the futures on the same manager?  This feels like something that maybe should live on a connection abstraction.  Or will it be impossible for python to end up in the state where you have one stream that's draining and one that's accepting new appends?

Reply:
It does seem odd to cancel everything sent later, though that's the semantics I copied from Pub/Sub. From what I can tell, `_reopen` isn't called until the old RPC has completely shut down.

https://github.com/googleapis/python-api-core/blob/8b65c93aef908c99784200b73ad270a0591481a8/google/api_core/bidi.py#L444-L451

Now that I look more closely, stream resumption only allows one request for the "initial request". 

https://github.com/googleapis/python-api-core/blob/8b65c93aef908c99784200b73ad270a0591481a8/google/api_core/bidi.py#L101-L105

That'll probably have to be refactored if/when we do stream resumption so that we can re-send any requests where we didn't get a response yet.